File: bench/python/bench.py
import time
import redis
import sys

PORT = 6381
HOST = "127.0.0.1"
TOTAL_OPS = 50000

def run():
    r = redis.Redis(host=HOST, port=PORT, db=0)
    
    print(f"Connecting to {HOST}:{PORT}...")
    try:
        r.set("warmup", "1")
        r.get("warmup")
    except Exception as e:
        print(f"Connection failed: {e}")
        sys.exit(1)

    print("Starting Python Benchmark...")
    start = time.time()

    for i in range(TOTAL_OPS):
        key = f"py:{i}"
        val = f"val:{i}"
        r.set(key, val)
        r.get(key)
        
    duration = time.time() - start
    ops = (TOTAL_OPS * 2) / duration
    print(f"Python (Sequential): {int(ops)} ops/sec")

    # --- Pipeline Benchmark ---
    print("Starting Python Benchmark (Pipelining batch=50)...")
    start = time.time()
    pipe = r.pipeline(transaction=False)
    
    BATCH_SIZE = 50
    for i in range(0, TOTAL_OPS, BATCH_SIZE):
        for j in range(BATCH_SIZE):
            if i + j >= TOTAL_OPS: break
            key = f"py:{i+j}"
            val = f"val:{i+j}"
            pipe.set(key, val)
            pipe.get(key)
        pipe.execute()
        
    duration = time.time() - start
    pipe_ops = (TOTAL_OPS * 2) / duration
    print(f"Python (Pipelined): {int(pipe_ops)} ops/sec")

if __name__ == "__main__":
    run()

File: cmd/bench/node/bench.js
const { createClient } = require('redis');

const PORT = 6381;
const HOST = '127.0.0.1';
const TOTAL_OPS = 50000;

async function run() {
    const client = createClient({
        url: `redis://${HOST}:${PORT}`
    });

    client.on('error', (err) => console.log('Redis Client Error', err));

    await client.connect();
    console.log(`Connected to ${HOST}:${PORT}`);

    // Warmup
    await client.set('warmup', '1');
    await client.get('warmup');

    console.log('Starting Node.js Benchmark (node-redis)...');
    const start = process.hrtime();

    for (let i = 0; i < TOTAL_OPS; i++) {
        await client.set(`key:${i}`, `value:${i}`);
        await client.get(`key:${i}`);
    }

    const end = process.hrtime(start);
    const duration = end[0] + end[1] / 1e9;
    const ops = (TOTAL_OPS * 2) / duration;

    console.log(`Node.js (Sequential): ${Math.round(ops)} ops/sec`);

    // --- Pipeline Benchmark ---
    console.log('Starting Node.js Benchmark (Pipelining batch=50)...');
    const pipeStart = process.hrtime();

    const BATCH_SIZE = 50;
    for (let i = 0; i < TOTAL_OPS; i += BATCH_SIZE) {
        const promises = [];
        for (let j = 0; j < BATCH_SIZE && i + j < TOTAL_OPS; j++) {
            promises.push(client.set(`key:${i + j}`, `val:${i + j}`));
            promises.push(client.get(`key:${i + j}`));
        }
        await Promise.all(promises);
    }

    const pipeEnd = process.hrtime(pipeStart);
    const pipeDuration = pipeEnd[0] + pipeEnd[1] / 1e9;
    const pipeOps = (TOTAL_OPS * 2) / pipeDuration;
    console.log(`Node.js (Pipelined): ${Math.round(pipeOps)} ops/sec`);

    await client.disconnect();
}

run().catch(console.error);
